File: python/src/n0542_01_matrix.py
from collections import deque


def update_matrix(matrix: list[list[int]]) -> list[list[int]]:
    m = len(matrix)
    n = len(matrix[0])
    result = [[-1 for _ in range(n)] for _ in range(m)]
    queue: deque[tuple[int, int]] = deque()
    for i in range(m):
        for j in range(n):
            if matrix[i][j] == 0:
                result[i][j] = 0
                queue.append((i, j))
    while queue:
        (i, j) = queue.popleft()
        if i > 0 and result[i - 1][j] == -1:
            result[i - 1][j] = result[i][j] + 1
            queue.append((i - 1, j))
        if j > 0 and result[i][j - 1] == -1:
            result[i][j - 1] = result[i][j] + 1
            queue.append((i, j - 1))
        if i + 1 < m and result[i + 1][j] == -1:
            result[i + 1][j] = result[i][j] + 1
            queue.append((i + 1, j))
        if j + 1 < n and result[i][j + 1] == -1:
            result[i][j + 1] = result[i][j] + 1
            queue.append((i, j + 1))
    return result

File: rust/src/n0542_01_matrix.rs
use std::collections::VecDeque;

pub fn update_matrix(matrix: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let (m, n) = (matrix.len(), matrix[0].len());
    let mut result: Vec<Vec<i32>> = vec![vec![-1; n]; m];
    let mut queue = VecDeque::new();
    for (i, row) in matrix.iter().enumerate() {
        for (j, &value) in row.iter().enumerate() {
            if value == 0 {
                result[i][j] = 0;
                queue.push_back((i, j));
            }
        }
    }
    while let Some((i, j)) = queue.pop_front() {
        if i > 0 && result[i - 1][j] == -1 {
            result[i - 1][j] = result[i][j] + 1;
            queue.push_back((i - 1, j));
        }
        if j > 0 && result[i][j - 1] == -1 {
            result[i][j - 1] = result[i][j] + 1;
            queue.push_back((i, j - 1));
        }
        if i + 1 < m && result[i + 1][j] == -1 {
            result[i + 1][j] = result[i][j] + 1;
            queue.push_back((i + 1, j));
        }
        if j + 1 < n && result[i][j + 1] == -1 {
            result[i][j + 1] = result[i][j] + 1;
            queue.push_back((i, j + 1));
        }
    }
    result
}
